### PR TITLE
Add HBase view for the addon recommender

### PR DESCRIPTION
**Pull request by Dexterp37**
This PR adds a scala job that scans the main summary and feeds addon plus some other data into an HBase instance.

**Do not review yet**.

### CODE REVIEW

File: src/main/scala/com/mozilla/telemetry/views/HBaseAddonRecommenderView.scala [hunk 1]
```diff
@@ -0,0 +1,154 @@
...
+
+  private[views] val tableName = "addon_recommender_view"
+  private[views] val columnFamily = "cf"
+  private[views] val column = "payload"
+
+  private[views] def createHBaseTable(useCompression: Boolean = true)(implicit config: HBaseConfig): Unit = {
+    val admin = new HBaseAdmin(config.get)
+    if (!admin.tableExists(tableName)) {
+      /*
+         About 200 GB of compressed data is generated per day. As we want to store data for 90 days we expect to
```
Comment (maurodoglio) [inline src/main/scala/com/mozilla/telemetry/views/HBaseAddonRecommenderView.scala]: Please delete this comment as it is obsolete. We should probably tune the region split value to reflect the kind of small dataset that we are storing.

File: src/main/scala/com/mozilla/telemetry/views/HBaseAddonRecommenderView.scala [hunk 2]
```diff
@@ -0,0 +1,154 @@
...
+
+        // Get the data for the desired date.
+        val data = datasetForDate(currentDateString)
+
+        // Only get the most recent entry (by subsession start date rather than submission
+        // date) for each client.
+        val clientsData: DataFrame = data
+            .select(
+              $"client_id",
+              $"document_id",
```
Comment (maurodoglio) [inline src/main/scala/com/mozilla/telemetry/views/HBaseAddonRecommenderView.scala]: We don't need document_id anymore, do we?

File: src/main/scala/com/mozilla/telemetry/views/HBaseAddonRecommenderView.scala [hunk 3]
```diff
@@ -0,0 +1,154 @@
...
+              $"scalar_parent_browser_engagement_unique_domains_count",
+              $"active_addons",
+              row_number()
+                .over(Window.partitionBy("client_id")
+                  .orderBy(desc("subsession_start_date")))
+                .alias("clientid_rank"))
+          .where($"clientid_rank" === 1)
+
+        // Convert the DataFrame to JSON and get an RDD out of it.
+        val subset = data.select("client_id", "document_id", "subsession_start_date")
```
Comment (maurodoglio) [inline src/main/scala/com/mozilla/telemetry/views/HBaseAddonRecommenderView.scala]: s/data/clientsData on this line and the one below.

File: src/main/scala/com/mozilla/telemetry/views/HBaseAddonRecommenderView.scala [hunk 4]
```diff
@@ -0,0 +1,154 @@
...
+                                   "locale",
+                                   "os",
+                                   "places_bookmarks_count",
+                                   "scalar_parent_browser_engagement_tab_open_event_count",
+                                   "scalar_parent_browser_engagement_total_uri_count",
+                                   "scalar_parent_browser_engagement_unique_domains_count",
+                                   "active_addons").toJSON.rdd
+
+        // Build an RDD containing (key, (document, timestamp)) tuples: one per client.
+        val rdd = subset.rdd.zip(jsonData).flatMap{ case (row, json) =>
```
Comment (maurodoglio) [inline src/main/scala/com/mozilla/telemetry/views/HBaseAddonRecommenderView.scala]: I'm wondering if there's a better way to obtain key/value tuples out of the rdd now that we only have client_id as the key

File: src/main/scala/com/mozilla/telemetry/views/HBaseAddonRecommenderView.scala [hunk 5]
```diff
@@ -0,0 +1,154 @@
...
+          val clientId = row.getString(0)
+          val startDate = row.getString(2)
```
Comment (maurodoglio) [inline src/main/scala/com/mozilla/telemetry/views/HBaseAddonRecommenderView.scala]: you can probably get rid of startDate & co now that we only keep one value per client. timestamp will be added automatically by hbase, so we don't need to handle that.

File: src/test/scala/com/mozilla/telemetry/views/HBaseAddonRecommenderViewTest.scala [hunk 6]
```diff
@@ -0,0 +1,103 @@
...
+
+import org.apache.hadoop.hbase.client.HBaseAdmin
+import org.apache.spark.sql.SparkSession
+import org.joda.time.DateTime
+import org.scalatest.{BeforeAndAfterAll, FlatSpec, Matchers}
+import org.scalatest.tagobjects.Slow
+import unicredit.spark.hbase._
+
+case class TestMainSummaryPing(client_id: Option[String],
+                               document_id: Option[String],
```
Comment (maurodoglio) [inline src/test/scala/com/mozilla/telemetry/views/HBaseAddonRecommenderViewTest.scala]: you can get rid of document_id

File: src/main/scala/com/mozilla/telemetry/views/HBaseAddonRecommenderView.scala (hunk 5 above)
Comment (Dexterp37) [inline src/main/scala/com/mozilla/telemetry/views/HBaseAddonRecommenderView.scala]: I removed it and it turns out that the code was implicitly performing data validation and discarding entries with invalid dates in the `subsession_start_date` field. I changed the code to keep the validation part and dropped adding it to the HBase.

**Pull request merged**